### PR TITLE
Add pundit

### PR DESCRIPTION
**Pull request by jessjno**
added pundit authorizations for comments, follow requests, photos, and users

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,12 +1,26 @@
 class ApplicationController < ActionController::Base
+  include Pundit
   before_action :authenticate_user!
-
+  
   before_action :configure_permitted_parameters, if: :devise_controller?
+  
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
 
+  # after_action :verify_authorized, unless: :devise_controller?
```
Comment (armstrca) [inline app/controllers/application_controller.rb]: Delete commented code

File: app/controllers/photos_controller.rb [hunk 2]
```diff
@@ -1,5 +1,7 @@
 class PhotosController < ApplicationController
   before_action :set_photo, only: %i[ show edit update destroy ]
+  before_action :ensure_current_user_is_owner, only: [:destroy, :update, :edit]
+  # before_action :ensure_user_is_authorized, only: [:show]
```
Comment (armstrca) [inline app/controllers/photos_controller.rb]: Delete commented code

File: app/controllers/photos_controller.rb [hunk 3]
```diff
@@ -67,4 +70,16 @@ def set_photo
...
       params.require(:photo).permit(:image, :comments_count, :likes_count, :caption, :owner_id)
     end
+
+    def ensure_current_user_is_owner
+      if current_user != @photo.owner
+        redirect_back fallback_location: root_url, alert: "You're not authorized for that."
+      end
+    end
+
+    # def ensure_user_is_authorized
```
Comment (armstrca) [inline app/controllers/photos_controller.rb]: Delete commented code

**Review by armstrca — COMMENTED** [review]
Alice is not able to delete her own comments

**Review by armstrca — CHANGES_REQUESTED** [review]
Please adjust for the issues mentioned

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,6 +1,30 @@
...
+  def followers
+    authorize(@user)
+  end
+
+  def following
+    authorize(@user)
+  end
+
+  def discover
+    authorize(@user)
```
Comment (hatcheta1) [inline app/controllers/users_controller.rb]: can DRY this up by defining a ```private``` method and using ```before_action```

File: app/policies/photo_policy.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  def update?
+    user == photo.owner ||
+    !photo.owner.private? ||
+    photo.owner.followers.include?(user)
+  end
+
+  def destroy?
+    user == photo.owner ||
+    !photo.owner.private? ||
+    photo.owner.followers.include?(user)
```
Comment (hatcheta1) [inline app/policies/photo_policy.rb]: Do you want the photo owner's followers to be able to delete their photos?

**Review by hatcheta1 — CHANGES_REQUESTED** [review]
See Calvin's comments about deleting commented-out code

File: app/controllers/photos_controller.rb [hunk 6]
```diff
@@ -67,4 +69,10 @@ def set_photo
     def photo_params
       params.require(:photo).permit(:image, :comments_count, :likes_count, :caption, :owner_id)
     end
+
+    def ensure_current_user_is_owner
```
Comment (Testell) [inline app/controllers/photos_controller.rb:73]: Good use of consistent/conventional naming conventions

File: app/controllers/comments_controller.rb [hunk 7]
```diff
@@ -67,4 +71,11 @@ def set_comment
     def comment_params
       params.require(:comment).permit(:author_id, :photo_id, :body)
     end
+
+    def is_an_authorized_user
```
Comment (Testell) [inline app/controllers/comments_controller.rb]: Nice job securing this down!

File: app/policies/user_policy.rb
Comment (Testell) [inline app/policies/user_policy.rb:1]: Nice consistent formatting

**Review by Testell — APPROVED** [review]

**Comment by armstrca** [comment]
Currently your handling for private profiles prevents `current_user` from being able to access the private user's profile at all, but it should really just be that you can't see the private user's photos (like how Instagram/the matchthetarget site work). 

Buuuuuuuuuuut that seems to be an error in the lesson itself, so I'm not going to dock anyone for it lol

The correct version would be something like:
```ruby
# app/policies/user_policy.rb
# ... other code
def show?
  true
end
# ... other code
```

and then preventing `current_user` from seeing the person's photos & pending requests using their respective policies.

File: app/controllers/comments_controller.rb [hunk 8]
```diff
@@ -1,5 +1,6 @@
 class CommentsController < ApplicationController
   before_action :set_comment, only: %i[ show edit update destroy ]
+  before_action :is_an_authorized_user, only: [:create]
```
Comment (armstrca) [inline app/controllers/comments_controller.rb:4]: If you add this to the top of your file (e.g. line 4):
```ruby
  before_action { authorize(@comment || Comment) }
```

You can remove all other instances of `authorize(@comment)`. This works for other controllers, too.

It's optional/not graded for this assignment, but you'll get more points on your final project if you go out of your way to follow DRY Principles like that.

File: app/policies/comment_policy.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  def edit?
+    comment.author_id == user.id
+  end
+
+  def update?
+    comment.author_id == user.id
+  end
+
+  def destroy?
+    user == comment.author
```
Comment (hatcheta1) [inline app/policies/comment_policy.rb:22]: I think the edit and update policies could be the same as this one and it's a bit more readable -- but that's not a necessary change

File: app/views/comments/_comment.html.erb
Comment (hatcheta1) [inline app/views/comments/_comment.html.erb:1]: nice job!

**Review by hatcheta1 — APPROVED** [review]
looks good!

File: app/policies/photo_policy.rb (hunk 5 above)
Comment (armstrca) [reply app/policies/photo_policy.rb]: Yeah, Ashanti's right - destroy, update, and edit policies should only be `user == photo.owner`

Currently you're handling this with conditionals in the view, which is good and correct, but leaving the policy like this is still a potential security vulnerability.

File: app/policies/comment_policy.rb (hunk 9 above)
Comment (armstrca) [reply app/policies/comment_policy.rb:22]: Agreed ^

File: app/policies/comment_policy.rb
Comment (armstrca) [inline app/policies/comment_policy.rb:1]: current_user does not seem to be able to create comments anywhere on the website

File: app/policies/follow_request_policy.rb
Comment (armstrca) [inline app/policies/follow_request_policy.rb:1]: I get this error any time I try to follow, un-follow, or un-request someone:

Screenshot 2024-11-05 103039

**Review by armstrca — CHANGES_REQUESTED** [review]

**Review by armstrca — APPROVED** [review]
gundam-mobile-suit-gundam 1

**Pull request merged**